tests/gbs/test_subgraph.py

```diff
@@ -108,195 +110,88 @@ def test_returns_densest(self, graph, monkeypatch):
-
-    def test_normal_conditions_shrink(self, graph):
-        """Test if function returns correct subgraph under normal conditions, where one needs to
-        shrink to a smaller subgraph. We consider the 7-node graph given by the below adjacency
-        matrix starting with a subgraph of the nodes [0, 1, 2, 3] and aiming to shrink to 3
-        nodes. We can see that there are four candidate nodes to remove (corresponding degree
-        shown in brackets): node 0 (degree 3), node 1 (degree 3), node 2 (degree 2) and node 4 (
-        degree 3). Hence, since node 2 has the lowest degree, we expect the returned solution
-        to be [0, 1, 3]."""
-        adj = (
-            (0, 1, 0, 0, 0, 1, 1),
-            (1, 0, 1, 1, 0, 0, 0),
-            (0, 1, 0, 1, 0, 0, 0),
-            (0, 1, 1, 0, 1, 0, 0),
-            (0, 0, 0, 1, 0, 0, 0),
-            (1, 0, 0, 0, 0, 0, 0),
-            (1, 0, 0, 0, 0, 0, 0),
-        )
-        graph = nx.Graph(0.5 * np.array(adj))  # multiply by 0.5 to follow weightings of adj fixture
-        s = subgraph.greedy_degree(subgraphs=[[0, 1, 2, 3]], graph=graph, target=3)
-        assert np.allclose(s, [0, 1, 3])
+    def test_input_not_subgraph(self):
+        """Test if function raises a ``ValueError`` when input is not a subgraph"""
+        dim = 5
+        with pytest.raises(ValueError, match="Input is not a valid subgraph"):
+            subgraph.resize([dim + 1], nx.complete_graph(dim))
+
+    def test_invalid_size(self):
+        """Test if function raises a ``ValueError`` when an invalid size is requested"""
+        dim = 5
+        with pytest.raises(ValueError, match="Requested sizes must be within size range of graph"):
+            subgraph.resize([0, 1], nx.complete_graph(dim), sizes=[-1, dim + 1])
+
+    @pytest.mark.parametrize("dim", [5, 6, 7])
+    def test_full_range(self, dim):
+        """Test if function correctly resizes to full range of sizes when the ``sizes`` argument
+        is not specified."""
+        g = nx.complete_graph(dim)
+        resized = subgraph.resize([0, 1, 2], g)
+
+        assert set(resized.keys()) == set(range(1, dim))
+
+        subgraph_sizes = [len(resized[i]) for i in range(1, dim)]
+
+        assert subgraph_sizes == list(range(1, dim))
+
+    def test_correct_resize(self):
+        """Test if function correctly resizes on a fixed example where the ideal resizing is
+        known. The example is a lollipop graph of 6 fully connected nodes and 2 nodes on the
+        lollipop stick. An edge between node zero and node ``dim - 1`` (the lollipop node
+        connecting to the stick) is removed and a starting subgraph of nodes ``[2, 3, 4, 5,
+        6]`` is selected. The objective is to add-on 1 and 2 nodes and remove 1 node."""
+        dim = 6
+        g = nx.lollipop_graph(dim, 2)
+        g.remove_edge(0, dim - 1)
+
+        s = [2, 3, 4, 5, 6]
+        sizes = [4, 6, 7]
+
+        ideal = {4: [2, 3, 4, 5], 6: [1, 2, 3, 4, 5, 6], 7: [0, 1, 2, 3, 4, 5, 6]}
```

Comment:
For the lollipop graph, all nodes in the "candy" have the same degree, so when choosing nodes to add/remove I would expect there to be a random breaking of ties. Then why does this test always return the "ideal" subgraph? For example, I would expect that for size 6, either [1, 2, 3, 4, 5, 6] or [0, 2, 3, 4, 5, 6] could show up. Am I missing something?

Reply:
The key is the removed edge between nodes 0 and 5. This makes node 0 have lower degree with respect to the subgraph, so we always pick node 1 to add in first.

When shrinking, we always should remove node 6 as it is a stick node.

Here's a plot:
![image](https://user-images.githubusercontent.com/49409390/66598395-f45a4f80-eb6e-11e9-9e7f-063123c6a774.png)

Reply:
Missed the removed node! Smart :thinking: :lollipop: